src/hotspot/share/gc/g1/heapRegion.inline.hpp

```diff
@@ -182,6 +182,27 @@ inline size_t HeapRegion::block_size(const HeapWord *addr) const {
   return block_size_using_bitmap(addr, G1CollectedHeap::heap()->concurrent_mark()->prev_mark_bitmap());
 }
 
+inline size_t HeapRegion::block_size_resolve(const HeapWord *addr) const {
+  if (addr == top()) {
+    return pointer_delta(end(), addr);
+  }
+
+  if (block_is_obj(addr)) {
+    oop obj = oop(addr);
+#ifdef _LP64
+#ifdef ASSERT
+    assert(UseCompactObjectHeaders && !G1CollectedHeap::heap()->collector_state()->in_full_gc(), "Illegal/excessive resolve during full-GC");
+#endif
+    if (obj->is_forwarded()) {
+      obj = obj->forwardee();
```

Comment:
1. I couldn't find the corresponding change upstream;
2. Is it possible the obj is in the middle of being forwarded?

Reply:
1. upstream made some changes which already avoid iterate forwarded objects.
2. Yes, it is possible.

Reply:
There're few places which might have forwarded oop. Should be all handled already.

Reply:
I see. So the obj here won't be forwarded because it should be in non-cset old region.